python/ray/tune/search/vizier/vizier_search.py

```diff
@@ -0,0 +1,246 @@
+            )
+        )
+        tuner.fit()
+
+    """
+
+    def __init__(
+        self,
+        space: Optional[Union[Dict, "svz.SearchSpace"] = None,
+        metric: Optional[str] = None,
+        mode: Optional[str] = None,
+        algorithm: Optional[str] = 'DEFAULT',
+    ):
+        assert IMPORT_SUCCESSFUL, 'Vizier must be installed with `pip install google-vizier[jax]`.'
+        super(VizierSearch, self).__init__(metric=metric, mode=mode)
+        self._algorithm = algorithm
+
+        # For Vizier to identify the unique study.
+        self._study_id = f'ray_vizier_{uuid.uuid1()}'
+
+        # Mapping from Ray trial id to Vizier Trial client.
+        self._active_trials: Dict[str, clients.Trial] = {}
+
+        # Vizier service client.
+        self._study_client: Optional[clients.Study] = None
+
+        if isinstance(space, dict) and space:
+            resolved_vars, domain_vars, grid_vars = parse_spec_vars(space)
+            if domain_vars or grid_vars:
+                logger.warning(
+                    UNRESOLVED_SEARCH_SPACE.format(par="space", cls=type(self))
+                )
+                space = self.convert_search_space(space)
+            self._space = space
+        elif isinstance(space, svz.SearchSpace):
+            self._space = space
+        else:
+            raise TypeError("Space must be Tune space or Vizier SearchSpace. Got {}.".format(type(space)))
+
+        if self._space:
+            self._setup_vizier()
+    
+    def set_search_properties(
+        self, metric: Optional[str], mode: Optional[str], config: Dict, **spec
+    ) -> bool:
+        if self._study_client:  # The study is already configured.
+            return False
+
+        self._metric = metric or DEFAULT_METRIC
+        self._mode = mode
+        self._space = vzr.SearchSpaceConverter.to_vizier(config)
+   
+        self._setup_vizier()
+        return True
+
+    def _setup_vizier(self) -> None:
+        if self._mode == 'max':
+            vizier_goal = svz.ObjectiveMetricGoal.MAXIMIZE
+        elif self._mode == 'min':
+            vizier_goal = svz.ObjectiveMetricGoal.MINIMIZE
+        else:
+            raise RuntimeError(UNDEFINED_METRIC_MODE.format(cls=self.__class__.__name__, metric=self._metric, mode = self._mode))
+
+        study_config = svz.StudyConfig(
+            search_space=self._space,
+            algorithm=self._algorithm,
+            metric_information=[svz.MetricInformation(self._metric, goal=vizier_goal)],
+        )
+        self._study_client = clients.Study.from_study_config(
+            study_config, owner='raytune', study_id=self._study_id
+        )
+
+    def on_trial_result(self, trial_id: str, result: Dict) -> None:
+        if trial_id not in self._active_trials:
+            raise RuntimeError(f'No active trial for {trial_id}')
+        if self._study_client is None:
+            raise RuntimeError(
+                'VizierSearch not initialized! Set a search space first.'
```

Comment:
Is this right? Why does this indicate that a search space is not set?

Reply:
Should be correct. `study_client` will be set whenever one of the methods (`__init__` or `set_search_properties`) is used, which requires knowing the search space in advance.